File: dotnet/src/MyDotey.SCF.Simple/Source/StringProperty/PropertieFile/PropertiesFileConfigurationSource.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

using NLog;
using Kajabity.Tools.Java;

namespace MyDotey.SCF.Source.StringProperty.PropertiesFile
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     */
    public class PropertiesFileConfigurationSource : StringPropertyConfigurationSource<PropertiesFileConfigurationSourceConfig>
    {
        private static Logger Logger = LogManager.GetCurrentClassLogger(typeof(PropertiesFileConfigurationSource));

        private JavaProperties _properties;

        public PropertiesFileConfigurationSource(PropertiesFileConfigurationSourceConfig config)
            : base(config)
        {
            _properties = new JavaProperties();

            try
            {
                using (Stream @is = new FileStream(config.FileName, FileMode.Open))
                {
                    if (@is == null)
                        Logger.Warn("file not found: {0}", config.FileName);

                    _properties.Load(@is, new UTF8Encoding(false));
                }
            }
            catch (Exception e)
            {
                Logger.Warn(e, "failed to load properties file: " + config.FileName);
            }
        }

        public override String GetPropertyValue(String key)
        {
            if (_properties == null)
                return null;

            return _properties.GetProperty(key);
        }
    }
}

File: java/src/main/java/org/mydotey/scf/source/stringproperty/propertiesfile/PropertiesFileConfigurationSource.java
package org.mydotey.scf.source.stringproperty.propertiesfile;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

import org.mydotey.scf.source.FileSourceType;
import org.mydotey.scf.source.stringproperty.StringPropertyConfigurationSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author koqizhao
 *
 * May 17, 2018
 */
public class PropertiesFileConfigurationSource
        extends StringPropertyConfigurationSource<PropertiesFileConfigurationSourceConfig> {

    private static final Logger LOGGER = LoggerFactory.getLogger(PropertiesFileConfigurationSource.class);

    private Properties _properties;

    public PropertiesFileConfigurationSource(PropertiesFileConfigurationSourceConfig config) {
        super(config);

        _properties = new Properties();
        InputStream is = null;
        try {
            if (config.getType() == FileSourceType.CLASS_PATH) {
                is = Thread.currentThread().getContextClassLoader()
                    .getResourceAsStream(config.getFileName());
            } else if(config.getType() == FileSourceType.LOCAL_SYSTEM) {
                is = new FileInputStream(config.getFileName());
            }

            if (is == null) {
                LOGGER.warn("file not found: {}", config);
                return;
            }

            _properties.load(is);
        } catch (Exception e) {
            LOGGER.warn("failed to load properties file: " + config, e);
        } finally {
            try {
                if (is != null)
                    is.close();
            } catch (IOException e) {

            }
        }
    }

    @Override
    public String getPropertyValue(String key) {
        if (_properties == null)
            return null;

        return _properties.getProperty(key);
    }

}
